telemetry/v1alpha1/telemetry.proto

```
@@ -0,0 +1,272 @@
+    HTTP = 1; 
+    // Selects for all non-HTTP traffic.
+    // NOTE: Tracing is currently only supported for HTTP. 
+    TCP = 2; 
+  }
+
+  // Optional. Specifies the intended direction of the traffic relative to the local proxy.
+  // Defaults to ALL if unset.
+  TrafficDirection traffic_direction = 1; 
+
+  // Optional. Specifies the protocol of the traffic.
+  // Defaults to ALL if unset.
+  Protocol protocol = 2; 
+
+  // Optional. The port on which the traffic is received.
+  // Defaults to ALL if unset.
+  PortSelector port = 3; 
+}
+
+// PortSelector specifies the number of a port to be used for
+// matching or selection for final routing.
+message PortSelector {
+  // Valid port number
+  uint32 number = 1;
+}
+
+// Used to bind Telemetry configuration to specific providers for
+// targeted customization.
+message ProviderRef {
+  // Required. Name of Telemetry provider in MeshConfig.
+  string name = 1;
+}
+
+// TracingConfig defines the workload-level overrides for tracing behavior within
+// a mesh. It can be used to enable/disable tracing, as well as to set sampling
+// rates and custom tag extraction.
+message TracingConfig {
+  // Optional. Name of providers to which this configuration should apply.
+  // If not specified, this will apply to the full list of known tracing
+  // providers.
+  repeated ProviderRef providers = 1;
+  
+  // Enables the tracing functionality. When this is set to `true`, the
+  // sidecar will report spans to a configured backend for all traffic with a trace
+  // context that specifies the trace is sampled. Additionally, spans will be
+  // generated for traffic without trace contexts based on the
+  // `sampler` configuration provided. If `report_spans` is `false` (or unset),
+  // the sidecar will ignore the incoming trace context, generating no spans for
+  // the traffic (the context will be silently forwarded). This is equivalent to 
+  // disabling tracing for the sidecar.
+  bool report_spans = 2; 
```

Why do we have a boolean on/off and a constantSampler=NEVER and percentage=0%? My understand is these all do the same, is that accurate?

---

`report_spans` controls whether or not tracing will be supported at all. It is a way to completely disable reporting of any spans regardless of trace context (it will silently pass on tags, but not report). This may not be necessary, but it was part of the prior Tracing API proposal, so I felt it should be carried over.

`constantSampler` and `percentageSampler` are only for traces that have no existing trace context (no sampling decision has already been made). They control the sampling rate for those requests. There is overlap between constantSampler=NEVER and a 0% sampler. We could drop the constant sampler here, if desired.